File: SeriesGuide/src/main/java/com/battlelancer/seriesguide/loaders/MovieLoader.java
/*
 * Copyright 2014 Uwe Trottmann
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.battlelancer.seriesguide.loaders;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import com.battlelancer.seriesguide.items.MovieDetails;
import com.battlelancer.seriesguide.util.DBUtils;
import com.battlelancer.seriesguide.util.MovieTools;
import com.uwetrottmann.androidutils.AndroidUtils;
import com.uwetrottmann.androidutils.GenericSimpleLoader;
import com.uwetrottmann.trakt.v2.entities.Ratings;
import org.joda.time.DateTime;

import static com.battlelancer.seriesguide.provider.SeriesGuideContract.Movies;

/**
 * Tries to load current movie details from trakt and TMDb, if failing tries to fall back to local
 * database copy.
 */
public class MovieLoader extends GenericSimpleLoader<MovieDetails> {

    private int mTmdbId;

    public MovieLoader(Context context, int tmdbId) {
        super(context);
        mTmdbId = tmdbId;
    }

    @Override
    public MovieDetails loadInBackground() {
        MovieDetails details = new MovieDetails();

        // try loading from trakt and tmdb
        if (AndroidUtils.isNetworkConnected(getContext())) {
            details = MovieTools.Download.getMovieDetails(getContext(), mTmdbId);
        }

        // update local database
        updateLocalMovie(getContext(), details, mTmdbId);

        // fill in details from local database
        Cursor movieQuery = getContext().getContentResolver()
                .query(Movies.buildMovieUri(mTmdbId), MovieQuery.PROJECTION, null, null, null);
        if (movieQuery == null || !movieQuery.moveToFirst() || movieQuery.getCount() < 1) {
            if (movieQuery != null) {
                movieQuery.close();
            }
            // ensure list flags are false on failure
            // (assumption: movie is not in db, it has the truth, so can't be in any lists)
            details.inCollection = false;
            details.inWatchlist = false;
            return details;
        }

        // set local state for watched, collected and watchlist status
        // assumption: local db has the truth for these
        details.inCollection = DBUtils.restoreBooleanFromInt(
                movieQuery.getInt(MovieQuery.IN_COLLECTION));
        details.inWatchlist = DBUtils.restoreBooleanFromInt(
                movieQuery.getInt(MovieQuery.IN_WATCHLIST));
        details.watched = DBUtils.restoreBooleanFromInt(movieQuery.getInt(MovieQuery.WATCHED));
        // also use local state of user rating
        details.userRating = movieQuery.getInt(MovieQuery.RATING_USER);

        // only overwrite other info if remote data failed to load
        if (details.released == null) {
            details.released = new DateTime(movieQuery.getLong(MovieQuery.RELEASED_UTC_MS));
        }
        if (details.traktRatings() == null) {
            details.traktRatings(new Ratings());
            details.traktRatings().rating = (double) movieQuery.getInt(MovieQuery.RATING_TRAKT);
            details.traktRatings().votes = movieQuery.getInt(MovieQuery.RATING_VOTES_TRAKT);
        }
        if (details.tmdbMovie() == null) {
            details.tmdbMovie(new com.uwetrottmann.tmdb.entities.Movie());
            details.tmdbMovie().imdb_id = movieQuery.getString(MovieQuery.IMDB_ID);
            details.tmdbMovie().title = movieQuery.getString(MovieQuery.TITLE);
            details.tmdbMovie().overview = movieQuery.getString(MovieQuery.OVERVIEW);
            details.tmdbMovie().poster_path = movieQuery.getString(MovieQuery.POSTER);
            details.tmdbMovie().runtime = movieQuery.getInt(MovieQuery.RUNTIME_MIN);
            details.tmdbMovie().vote_average = movieQuery.getDouble(MovieQuery.RATING_TMDB);
        }

        // clean up
        movieQuery.close();

        return details;
    }

    private static void updateLocalMovie(Context context,
            MovieDetails details, int tmdbId) {
        ContentValues values = MovieTools.buildBasicMovieContentValues(details);
        if (values.size() == 0) {
            // nothing to update, downloading probably failed :(
            return;
        }

        // if movie does not exist in database, will do nothing
        context.getContentResolver().update(Movies.buildMovieUri(tmdbId), values, null, null);
    }

    private interface MovieQuery {

        public String[] PROJECTION = {
                Movies.TITLE,
                Movies.OVERVIEW,
                Movies.RELEASED_UTC_MS,
                Movies.POSTER,
                Movies.WATCHED,
                Movies.IN_COLLECTION,
                Movies.IN_WATCHLIST,
                Movies.IMDB_ID,
                Movies.RUNTIME_MIN,
                Movies.RATING_TMDB,
                Movies.RATING_TRAKT,
                Movies.RATING_VOTES_TRAKT,
                Movies.RATING_USER
        };

        int TITLE = 0;
        int OVERVIEW = 1;
        int RELEASED_UTC_MS = 2;
        int POSTER = 3;
        int WATCHED = 4;
        int IN_COLLECTION = 5;
        int IN_WATCHLIST = 6;
        int IMDB_ID = 7;
        int RUNTIME_MIN = 8;
        int RATING_TMDB = 9;
        int RATING_TRAKT = 10;
        int RATING_VOTES_TRAKT = 11;
        int RATING_USER = 12;
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/movies/MovieLoader.kt
// SPDX-License-Identifier: GPL-3.0-or-later
// SPDX-FileCopyrightText: Copyright © 2020 Uwe Trottmann <uwe@uwetrottmann.com>

package com.battlelancer.seriesguide.movies

import android.content.Context
import com.battlelancer.seriesguide.R
import com.battlelancer.seriesguide.SgApp.Companion.getServicesComponent
import com.battlelancer.seriesguide.backend.settings.HexagonSettings
import com.battlelancer.seriesguide.movies.database.SgMovie
import com.battlelancer.seriesguide.movies.details.UiMovieDetails
import com.battlelancer.seriesguide.movies.tools.MovieDetails
import com.battlelancer.seriesguide.movies.tools.MovieDownloader.MovieDetailsResult
import com.battlelancer.seriesguide.movies.tools.MovieTools
import com.battlelancer.seriesguide.provider.SgRoomDatabase.Companion.getInstance
import com.battlelancer.seriesguide.tmdbapi.TmdbTools
import com.battlelancer.seriesguide.traktapi.TraktCredentials
import com.battlelancer.seriesguide.traktapi.TraktTools
import com.battlelancer.seriesguide.util.ImageTools
import com.battlelancer.seriesguide.util.RatingsTools
import com.battlelancer.seriesguide.util.TextTools
import com.battlelancer.seriesguide.util.TimeTools
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.threeten.bp.Instant
import org.threeten.bp.ZonedDateTime

/**
 * Tries to load current movie details from Trakt and TMDB, on failure tries to fall back to a local
 * database copy.
 */
class MovieLoader(
    private val context: Context,
    private val tmdbId: Int
) {

    sealed interface Result {
        data class Success(val details: UiMovieDetails) : Result
        object Error : Result
    }

    suspend fun loadInBackground(): Result = withContext(Dispatchers.Default) {
        load()
    }

    private suspend fun load(): Result {
        val movieTools = getServicesComponent(context).movieTools()

        // try loading from trakt and tmdb, this might return a cached response
        val detailsResult =
            movieTools.downloader
                .getMovieDetailsWithDefaults(tmdbId, true)

        val details: MovieDetails? = when (detailsResult) {
            is MovieDetailsResult.Success -> detailsResult.movieDetails
            is MovieDetailsResult.Error -> null
        }

        if (details != null) {
            // Update local database (no-op if movie not in database).
            movieTools.updateMovieWithTmdbId(tmdbId, details)
        }

        // Fill in or use cached details from local database
        val dbMovieOrNull = getInstance(context)
            .movieHelper()
            .getMovie(tmdbId)

        // Need at least details from either TMDB or the database
        if (details == null && dbMovieOrNull == null) {
            return Result.Error
        }

        return Result.Success(
            mapToUiMovieDetails(
                tmdbId,
                details,
                dbMovieOrNull,
                context
            )
        )
    }

    /**
     * Assumes at least one of [movieDetails] or [dbMovie] is not null.
     */
    fun mapToUiMovieDetails(
        tmdbId: Int,
        movieDetails: MovieDetails?,
        dbMovie: SgMovie?,
        context: Context
    ): UiMovieDetails {
        val tmdbMovie = movieDetails?.tmdbMovie

        val title = tmdbMovie?.title ?: dbMovie?.title

        // Metacritic only has English titles so mostly English-speaking users will use it,
        // so it's likely the original language of the movie is English.
        val titleForMetacritic = if (tmdbMovie?.original_language == "en") {
            tmdbMovie.original_title
        } else title

        val isWatched = dbMovie?.watchedOrDefault ?: false
        val plays = dbMovie?.playsOrDefault ?: 0

        // Release date and running time
        val runningTime = tmdbMovie?.runtime ?: dbMovie?.runtimeMinOrDefault
        val releaseDate = tmdbMovie?.release_date
            ?: dbMovie
                ?.let { MovieTools.movieReleaseDateFrom(it.releasedMsOrDefault) }
        val releaseDateAndRunningTime = TextTools.dotSeparate(
            releaseDate?.let { TimeTools.formatToLocalDate(context, it) },
            runningTime?.let { TimeTools.formatToHoursAndMinutes(context.resources, it) }
        )

        val lastUpdatedMillis = dbMovie?.lastUpdated ?: 0

        // hide check-in if not connected to trakt or hexagon is enabled
        val isConnectedToTrakt = TraktCredentials.get(context).hasCredentials()
        val hideCheckIn = !isConnectedToTrakt || HexagonSettings.isEnabled(context)

        val collection = tmdbMovie?.belongs_to_collection

        // Genres
        val genres = tmdbMovie?.genres?.let { TmdbTools.buildGenresString(it) }
            ?: dbMovie?.genres
        val genresOrUnknown = if (genres.isNullOrEmpty()) {
            context.getString(R.string.unknown)
        } else genres

        // Poster URLs
        val posterPath = tmdbMovie?.poster_path ?: dbMovie?.poster
        val hasPosterPath = !posterPath.isNullOrEmpty()
        val posterSmallSizeImageUrl: String?
        val posterOriginalSizeImageUrl: String?
        if (hasPosterPath) {
            posterSmallSizeImageUrl =
                TmdbTools.buildLargePosterUrl(context, posterPath)
                    .let { ImageTools.buildImageCacheUrl(it) }
            posterOriginalSizeImageUrl =
                TmdbTools.buildOriginalSizeImageUrl(context, posterPath)
                    .let { ImageTools.buildImageCacheUrl(it) }
        } else {
            posterSmallSizeImageUrl = null
            posterOriginalSizeImageUrl = null
        }

        val traktRatings = movieDetails?.traktRatings

        return UiMovieDetails(
            imdbId = tmdbMovie?.imdb_id ?: dbMovie?.imdbId,
            title = title,
            titleForMetacritic = titleForMetacritic,
            // No need to set no translation available message if empty, movie downloader does
            overview = tmdbMovie?.overview ?: dbMovie?.overview,
            // Ensure lists, watched flag are false and plays 0 if db movie is not in database
            inCollection = dbMovie?.inCollectionOrDefault ?: false,
            inWatchlist = dbMovie?.inWatchlistOrDefault ?: false,
            watched = isWatched,
            plays = plays,
            releaseDate = releaseDate,
            releaseDateAndRunningTime = releaseDateAndRunningTime,
            lastUpdatedText = TimeTools.formatToLocalDateAndTime(context, lastUpdatedMillis),
            isShareButtonEnabled = title != null,
            // Hide create event button if release date is yesterday or older
            isCalendarButtonGone = releaseDate == null
                    || Instant.ofEpochMilli(releaseDate.time)
                .isBefore(ZonedDateTime.now().minusDays(1).toInstant()),
            isCheckInButtonGone = hideCheckIn,
            watchedButtonText = TextTools.getWatchedButtonText(context, isWatched, plays),
            tmdbCollectionId = collection?.id,
            tmdbCollectionName = collection?.name,
            tmdbRating = RatingsTools.buildRatingString(
                tmdbMovie?.vote_average ?: dbMovie?.ratingTmdb
            ),
            tmdbVotes = RatingsTools.buildRatingVotesString(
                context,
                tmdbMovie?.vote_count ?: dbMovie?.ratingVotesTmdb
            ),
            traktRating = RatingsTools.buildRatingString(
                traktRatings?.rating ?: dbMovie?.ratingTrakt?.toDouble()
            ),
            traktVotes = RatingsTools.buildRatingVotesString(
                context,
                traktRatings?.votes ?: dbMovie?.ratingVotesTrakt
            ),
            userRating = dbMovie?.ratingUser,
            userRatingText = TraktTools.buildUserRatingString(context, dbMovie?.ratingUser),
            genres = genresOrUnknown,
            tmdbUrl = TmdbTools.buildMovieUrl(tmdbId),
            traktUrl = TraktTools.buildMovieUrl(tmdbId),
            posterSmallSizeImageUrl = posterSmallSizeImageUrl,
            posterOriginalSizeImageUrl = posterOriginalSizeImageUrl
        )
    }
}